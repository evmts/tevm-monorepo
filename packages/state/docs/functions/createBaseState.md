[**@tevm/state**](../README.md) • **Docs**

***

[@tevm/state](../globals.md) / createBaseState

# Function: createBaseState()

`Internal`

> **createBaseState**(`options`): [`BaseState`](../type-aliases/BaseState.md)

Creates the core data structure for ethereum state

## Parameters

• **options**: [`StateOptions`](../type-aliases/StateOptions.md)

## Returns

[`BaseState`](../type-aliases/BaseState.md)

Creates the core data structure the state manager operates on

## Source

[packages/state/src/createBaseState.js:20](https://github.com/evmts/tevm-monorepo/blob/main/packages/state/src/createBaseState.js#L20)
