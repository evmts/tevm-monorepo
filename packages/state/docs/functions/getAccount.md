[**@tevm/state**](../README.md) • **Docs**

***

[@tevm/state](../globals.md) / getAccount

# Function: getAccount()

> **getAccount**(`baseState`): (`address`) => `Promise`\<`undefined` \| `Account`\>

Gets the code corresponding to the provided `address`.

## Parameters

• **baseState**: [`BaseState`](../type-aliases/BaseState.md)

## Returns

`Function`

### Parameters

• **address**: `Address`

### Returns

`Promise`\<`undefined` \| `Account`\>

## Source

[packages/state/src/actions/getAccount.js:8](https://github.com/evmts/tevm-monorepo/blob/main/packages/state/src/actions/getAccount.js#L8)
