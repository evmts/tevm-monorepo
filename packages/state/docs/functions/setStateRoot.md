[**@tevm/state**](../README.md) • **Docs**

***

[@tevm/state](../globals.md) / setStateRoot

# Function: setStateRoot()

> **setStateRoot**(`baseState`): (`stateRoot`, `clearCache`?) => `Promise`\<`void`\>

Changes the currently loaded state root

## Parameters

• **baseState**: [`BaseState`](../type-aliases/BaseState.md)

## Returns

`Function`

### Parameters

• **stateRoot**: `Uint8Array`

• **clearCache?**: `boolean`

### Returns

`Promise`\<`void`\>

## Source

[packages/state/src/actions/setStateRoot.js:23](https://github.com/evmts/tevm-monorepo/blob/main/packages/state/src/actions/setStateRoot.js#L23)
