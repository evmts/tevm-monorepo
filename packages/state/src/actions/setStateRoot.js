import { bytesToHex } from 'viem'
import { generateCanonicalGenesis } from './generateCannonicalGenesis.js'

/**
 * Error thrown if state root doesn't exist
 */
export class NoStateRootExistsError extends Error {
	/**
	 * @override
	 * @type {'NoStateRootExistsError'}
	 */
	name = 'NoStateRootExistsError'
	/**
	 * @type {'NoStateRootExistsError'}
	 */
	_tag = 'NoStateRootExistsError'
}

/**
 * Changes the currently loaded state root
 * @type {import("../state-types/index.js").StateAction<'setStateRoot'>}
 */
export const setStateRoot = (baseState) => async (root) => {
	if (
		baseState.caches.accounts._checkpoints > 0 ||
		baseState.caches.storage._checkpoints > 0 ||
		baseState.caches.contracts._checkpoints > 0
	) {
		throw new Error('Attempted to setStateRoot with uncommitted checkpoints')
	}
	const genesis = baseState.stateRoots.get(bytesToHex(root))
	// TODO we need to handle the fok case here
	if (!genesis) {
		throw new NoStateRootExistsError(`State root for ${bytesToHex(root)} does not exist`)
	}
	const oldStateRoot = baseState.getCurrentStateRoot()
	try {
		baseState.setCurrentStateRoot(bytesToHex(root))
		await generateCanonicalGenesis(baseState)(genesis)
		baseState.logger.debug({ oldStateRoot, newStateRoot: root }, 'state root changed')
		return
	} catch (e) {
		baseState.setCurrentStateRoot(oldStateRoot)
		throw e
	}
}
