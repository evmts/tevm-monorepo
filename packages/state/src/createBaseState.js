import { AccountCache, CacheType, StorageCache } from '@ethereumjs/statemanager'
import { createLogger } from '@tevm/logger'
import { ContractCache } from './ContractCache.js'
import { checkpoint } from './actions/checkpoint.js'
import { commit } from './actions/commit.js'
import { generateCanonicalGenesis } from './actions/generateCannonicalGenesis.js'

/**
 * @type {import('viem').Hex}
 */
const INITIAL_STATE_ROOT = '0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421'

/**
 * @internal
 * Creates the core data structure for ethereum state
 * @param {import('./state-types/index.js').StateOptions} options
 * @returns {import('./BaseState.js').BaseState}
 * Creates the core data structure the state manager operates on
 */
export const createBaseState = (options) => {
	const logger = createLogger({
		level: options.loggingLevel ?? 'warn',
		name: '@tevm/state-manager',
	})
	/**
	 * @type {import('./state-types/StateRoots.js').StateRoots}
	 */
	const stateRoots = options.stateRoots ?? new Map()
	if (options.genesisState && options.currentStateRoot) {
		stateRoots.set(options.currentStateRoot, options.genesisState)
	} else {
		stateRoots.set(INITIAL_STATE_ROOT, options.genesisState ?? {})
	}
	let currentStateRoot = options.currentStateRoot ?? INITIAL_STATE_ROOT
	/**
	 * @type {import('./BaseState.js').BaseState}
	 */
	const state = {
		logger,
		getCurrentStateRoot: () => currentStateRoot,
		setCurrentStateRoot: (root) => {
			if (!state.stateRoots.has(root)) {
				throw new Error('Cannot set state root to non existing state root')
			}
			currentStateRoot = root
		},
		stateRoots,
		options,
		caches: {
			contracts: new ContractCache(),
			accounts: new AccountCache({
				size: 100_000,
				type: CacheType.ORDERED_MAP,
			}),
			storage: new StorageCache({
				size: 100_000,
				type: CacheType.ORDERED_MAP,
			}),
		},
		ready: () => genesisPromise.then(() => true),
	}
	const genesisPromise = (
		options.genesisState !== undefined && options.currentStateRoot === undefined
			? generateCanonicalGenesis(state)(options.genesisState)
			: Promise.resolve().then(() => {
					if (options.currentStateRoot) {
						state.setCurrentStateRoot(options.currentStateRoot)
						return generateCanonicalGenesis(state)(options.genesisState ?? stateRoots.get(options.currentStateRoot))
					}
					return Promise.resolve()
				})
	).then(async () => {
		await checkpoint(state)()
		await commit(state)()
		logger.debug('StateManager is ready')
	})

	return state
}
