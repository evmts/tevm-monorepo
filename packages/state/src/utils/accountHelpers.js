import { Rlp } from '@tevm/rlp'
import { createAccount, createAddressFromString as createAddress } from '@tevm/utils'

/**
 * Helper to create an address from a string
 * @param {string} addressString - The address string
 * @returns {import('@tevm/utils').EthjsAddress} The address
 */
export function createAddressFromString(addressString) {
	return createAddress(addressString)
}

/**
 * Creates an account from RLP serialized data
 * @param {Uint8Array} serialized - The RLP serialized account data
 * @returns {import('@tevm/utils').EthjsAccount} The account
 */
export function fromRlpSerializedAccount(serialized) {
	const decoded = Rlp.decode(serialized)
	if (!Array.isArray(decoded) || decoded.length !== 4) {
		throw new Error('Invalid RLP serialized account')
	}

	const [nonce, balance, storageRoot, codeHash] = decoded

	/** @type {{nonce: any, balance: any, storageRoot: any, codeHash: any}} */
	const accountData = {
		nonce,
		balance,
		storageRoot,
		codeHash,
	}

	return createAccount(accountData)
}

/**
 * Creates an account from account data
 * @param {{nonce?: bigint, balance?: bigint, storageRoot?: Uint8Array, codeHash?: Uint8Array}} accountData - The account data
 * @returns {import('@tevm/utils').EthjsAccount} The account
 */
export function fromAccountData(accountData) {
	return createAccount(accountData)
}
