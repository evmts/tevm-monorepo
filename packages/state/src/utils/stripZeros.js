import { InternalError } from '@tevm/errors'

/**
 * @param {Uint8Array} bytes
 * @returns {Uint8Array}
 */
export const stripZeros = (bytes) => {
	if (!(bytes instanceof Uint8Array)) {
		throw new InternalError('Unexpected type')
	}
	return bytes.slice(bytes.findIndex(/** @param {number} entry*/ (entry) => entry !== 0))
}
