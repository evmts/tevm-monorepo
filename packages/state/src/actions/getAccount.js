import { EthjsAccount } from '@tevm/utils'
import { getAccountFromProvider } from './getAccountFromProvider.js'

/**
 * Gets the code corresponding to the provided `address`.
 * @type {import("../state-types/index.js").StateAction<'getAccount'>}
 */
export const getAccount = (baseState) => async (address) => {
	const {
		caches: { accounts },
	} = baseState
	const elem = accounts.get(address)
	if (elem !== undefined) {
		return elem.accountRLP !== undefined ? EthjsAccount.fromRlpSerializedAccount(elem.accountRLP) : undefined
	}
	if (!baseState.options.fork?.url && elem === undefined) {
		return undefined
	}
	baseState.logger.debug({ address }, 'fetching account from remote RPC')
	const account = await getAccountFromProvider(baseState)(
		/** @type {import('@tevm/utils').Address}*/ (address.toString()),
	)
	if (
		account.nonce === 0n &&
		account.balance === 0n &&
		account.codeHash.every((d) => d === 0) &&
		account.storageRoot.every((d) => d === 0)
	) {
		return undefined
	}
	accounts.put(address, account)
	baseState.logger.debug({ address, account }, 'Cached forked account in state manager')
	return account
}
