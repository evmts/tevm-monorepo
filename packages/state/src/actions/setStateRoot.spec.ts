import { describe, expect, it } from 'bun:test'
import { EthjsAddress, bytesToHex } from '@tevm/utils'
import { hexToBytes } from 'ethereum-cryptography/utils'
import { createBaseState } from '../createBaseState.js'
import { getAccount } from './getAccount.js'
import { NoStateRootExistsError, setStateRoot } from './setStateRoot.js'

describe(setStateRoot.name, () => {
	it('should set state root', async () => {
		const address = `0x${'01'.repeat(20)}` as const
		const baseState = createBaseState({
			loggingLevel: 'warn',
		})
		await baseState.ready()
		const root = hexToBytes(`0x${'11'.repeat(32)}`)
		baseState.stateRoots.set(bytesToHex(root), {
			[address]: {
				balance: 420n,
				nonce: 1n,
				storageRoot: '0x69',
				codeHash: '0x420',
			},
		})
		await setStateRoot(baseState)(root)
		const account = await getAccount(baseState)(EthjsAddress.fromString(address))
		expect(account?.balance).toBe(420n)
	})

	it('should throw if no state root exists', async () => {
		const baseState = createBaseState({
			loggingLevel: 'warn',
		})
		await baseState.ready()
		const root = hexToBytes(`0x${'11'.repeat(32)}`)
		expect(() => setStateRoot(baseState)(root)).toThrowError(
			new NoStateRootExistsError(`State root for ${bytesToHex(root)} does not exist`),
		)
	})
})
