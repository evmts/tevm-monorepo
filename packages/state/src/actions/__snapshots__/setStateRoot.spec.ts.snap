// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`setStateRoot > should throw if no state root exists 1`] = `[NoStateRootExistsError: State root for 0x1111111111111111111111111111111111111111111111111111111111111111 does not exist]`;
